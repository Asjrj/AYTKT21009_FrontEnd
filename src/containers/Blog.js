import React from 'react'
import Notification from '../components/Notification'
import { likeBlog, deleteBlog, commentBlog } from '../reducers/blogReducer'
import { notify } from '../reducers/notificationReducer'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

class Blog extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      newComment: ''
    }
  }

  handleCommentChange = (event) => {
    event.preventDefault()
    this.setState({ newComment: event.target.value })
  }

  likeThisBlog = async (event) => {
    event.preventDefault()
    let blog = this.props.blogs.find((element) => {
      return element.id === event.target.getAttribute("id")
    })
    blog.likes++
    try {
      await this.props.likeBlog(blog, this.props.theUser.token)
    }
    catch (exception) {
      this.props.notify('Error performing like functionality', 'error', 5)
    }
  }

  deleteThisBlog = async (event) => {
    event.preventDefault()
    let delId = event.target.getAttribute("id")
    let deleteBlog = this.props.blogs.find((element) => {
      return element.id === delId
    })
    try {
      if (window.confirm(`delete ${deleteBlog.title} by ${deleteBlog.author}`)) {        
        this.props.deleteBlog(deleteBlog, this.props.theUser.token)
        window.location.replace('/');
      }
    }
    catch (exception) {
      this.props.notify('Error deleting blog', 'error', 5)
    }
  }

  commentThisBlog = async (event) => {
    event.preventDefault()
    let blogId = event.target.getAttribute("id")
    let blog = this.props.blogs.find((element) => {
      return element.id === blogId
    })
    try {
      if (this.state.newComment !== undefined && this.state.newComment !== '') {
        let comment = this.state.newComment
        await this.props.commentBlog(blog, comment)
        this.setState({ newComment: '' })
        this.props.notify(`comment: "${comment}" added`, 'info', 5)
      }
    }
    catch (exception) {
      this.props.notify('Error commenting blog', 'error', 5)
    }
  }

  render() {
    let blog = this.props.blogs.find(a => a.id === this.props.blogId)
    if (blog.comments === undefined) blog.comments = []
    let showDeleteButton = { display: '' }
    let userName = null
    if (blog && blog.user) {
      userName = blog.user.name
      if (this.props.theUser && this.props.theUser.name !== blog.user.name) {
        showDeleteButton = { display: 'none' }
      }
    }
    let showUser = { display: userName ? '' : 'none' }

    if (blog) {
      return (
        <div>
          <Notification />
          <h2>{blog.title}</h2>
          <div className='blogDetail'>
            <a href={blog.url}>{blog.url}</a><br />
            <a>{blog.likes} likes <button id={blog.id} onClick={this.likeThisBlog}>like</button></a><br />
            <a style={showUser}>added by {userName}</a><br />
            <button style={showDeleteButton} id={blog.id} onClick={this.deleteThisBlog}>delete</button>
          </div>
          <div>
            <h3>Comments</h3>
            <ul>
              {blog.comments.map(comment => <li key={comment}>{comment}</li>)}
            </ul>
          </div>
          <form >
            <div>
              <input type="text" value={this.state.newComment} onChange={this.handleCommentChange} />
              <button id={blog.id} onClick={this.commentThisBlog}>add comment</button>
            </div>
          </form>
        </div>
      )
    }
    else {
      return (<div></div>)
    }
  }
}

Blog.propTypes = {
  blogId: PropTypes.string.isRequired
}
const mapStateToProps = (state) => {
  return {
    blogs: state.blogs.blogs,
    theUser: state.users.user
  }
}
const mapDispatchToProps = {
  notify,
  likeBlog,
  deleteBlog,
  commentBlog
}
export default connect(mapStateToProps, mapDispatchToProps)(Blog)
