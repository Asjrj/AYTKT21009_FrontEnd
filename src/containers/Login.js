import React from 'react'
import loginService from '../services/login'
import LoginForm from '../components/LoginForm'
import Notification from '../components/Notification'
import { notify } from '../reducers/notificationReducer'
import { setUser } from '../reducers/userReducer'
import { connect } from 'react-redux'

class Login extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      username: '',
      password: ''
    }
  }

  handleUsernameChange = (event) => {
    this.setState({ username: event.target.value })
  }

  handlePasswordChange = (event) => {
    this.setState({ password: event.target.value })
  }

  handleLogin = async (event) => {
    event.preventDefault()
    try {
      const user = await loginService.login({
        username: this.state.username,
        password: this.state.password
      })
      this.props.setUser(user)
    }
    catch (exception) {
      this.props.notify('Invalid username or password', 'error', 5)
    }
  }

  render() {
    return (
      <div>
        <h2>Log in to application</h2>
        <Notification />
        <LoginForm
          username={this.state.username}
          password={this.state.password}
          handleUsernameChange={this.handleUsernameChange}
          handlePasswordChange={this.handlePasswordChange}
          login={this.handleLogin}
        ></LoginForm>
      </div >
    )
  }
}

const mapStateToProps = (state) => {
  return {
    theUser: state.users.user
  }
}
const mapDispatchToProps = {
  notify,
  setUser
}
export default connect(mapStateToProps, mapDispatchToProps)(Login)