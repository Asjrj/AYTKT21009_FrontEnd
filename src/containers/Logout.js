import React from 'react'
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom'


class Logout extends React.Component {

  logout = (event) => {
    event.preventDefault()
    window.localStorage.removeItem('blogUser')
    window.location.replace('/');
  }

  render() {
    const menuStyle = {
      fontFamily: 'Arial, Verdana',
      color: 'green',
      fontSize: 16,
      marginBottom: 10,
      padding: 10
    }
    const activeMenuStyle = {
      fontFamily: 'Arial, Verdana',
      fontWeight: 'bold',
      color: 'dark-green',
      fontSize: 16,
    }
    return (
      <div>
        <h2>Blogs</h2>
        <form onSubmit={this.logout}>
          <div style={menuStyle}>
            <NavLink exact to="/" activeStyle={activeMenuStyle}>Blogs</NavLink>&nbsp;
            <NavLink exact to="/users" activeStyle={activeMenuStyle}>Users</NavLink>&nbsp;
            <a>{this.props.user.name} logged in &nbsp;</a> <button type="submit">logout</button>
          </div>
        </form>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    theUser: state.users.user
  }
}
export default connect(mapStateToProps)(Logout)
