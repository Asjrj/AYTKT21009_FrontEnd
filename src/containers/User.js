import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { ListGroup, ListGroupItem } from 'react-bootstrap'

export class User extends React.Component {

  logout = (event) => {
    event.preventDefault()
    window.localStorage.removeItem('blogUser')
    window.location.reload()
  }

  getTheBlogs = (userId) => {
    let theBlogs = []
    this.props.blogs.forEach(element => {
      if (element.user && element.user._id === userId) {
        theBlogs.push(element)
      }
    })
    return (theBlogs)
  }

  render() {
    let theUser = this.props.users.find(a => a.id === this.props.userId)
    return (
      <div>
        <h2>{theUser.name}</h2>
        <h3>Added blogs</h3>
        <ListGroup>
          {this.getTheBlogs(this.props.userId).map(blog => <ListGroupItem key={blog.id}>{blog.title} by {blog.author}</ListGroupItem>)}
        </ListGroup>
      </div>
    )
  }
}

User.propTypes = {
  blogs: PropTypes.array.isRequired
}
const mapStateToProps = (state, ownProps) => {
  return {
    blogs: state.blogs.blogs,
    users: state.users.users,
    userId: ownProps.theUserId
  }
}
export default connect(mapStateToProps)(User)
