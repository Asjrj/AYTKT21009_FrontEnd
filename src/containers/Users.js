import React from 'react'
import { Link } from 'react-router-dom'

class Users extends React.Component {

  logout = (event) => {
    event.preventDefault()
    window.localStorage.removeItem('blogUser')
    window.location.reload()
  }

  render() {
    return (
      <div>
        <h2>Blogs</h2>
        <form onSubmit={this.logout}>
          <p>{this.props.user.name} logged in&nbsp; <button type="submit">logout</button>
          </p>
        </form>
        <h3>Users</h3>
        <table>
          <tbody>
            <tr><td></td><td>blogs added</td></tr>
            {this.props.users
              .map(user =>
                <tr key={user.id}>
                  <td><Link to={`/users/${user.id}`}>{user.name}</Link></td><td>{user.blogs.length}</td>
                </tr>
              )}
          </tbody>
        </table>
      </div>
    )
  }
}

export default Users
