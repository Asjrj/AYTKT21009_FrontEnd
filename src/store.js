import { createStore, combineReducers, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import notificationReducer from './reducers/notificationReducer'
import userReducer from './reducers/userReducer'

const reducer = combineReducers({
  users: userReducer,
  notification: notificationReducer
})

const store = createStore(reducer, applyMiddleware(thunk))

export default store