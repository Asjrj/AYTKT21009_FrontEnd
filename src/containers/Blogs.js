import React from 'react'
import blogService from '../services/blogs'
import Blog from '../components/Blog'
import BlogForm from '../components/BlogForm'
import Notification from '../components/Notification'
import Togglable from '../components/Togglable'
import PropTypes from 'prop-types'
import { notify } from '../reducers/notificationReducer'
import { connect } from 'react-redux'


class Blogs extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      newTitle: '',
      newAuthor: '',
      newUrl: ''
    }
  }

  handleTitleChange = () => {
    return (event) => {
      event.preventDefault()
      this.setState({ newTitle: event.target.value })
    }
  }

  handleAuthorChange = () => {
    return (event) => {
      event.preventDefault()
      this.setState({ newAuthor: event.target.value })
    }
  }

  handleUrlChange = () => {
    return (event) => {
      event.preventDefault()
      this.setState({ newUrl: event.target.value })
    }
  }

  handleCreate = () => {
    return async (event) => {
      event.preventDefault()
      if (this.state.newTitle === '' || this.state.newAuthor === '' || this.state.newUrl === '') {        
        this.props.notify('Title, author and url must be given', 'error', 5)
        return
      }
      const newBlog = {
        id: "123",
        title: this.state.newTitle,
        author: this.state.newAuthor,
        url: this.state.newUrl
      }
      try {
        let addedBlog = await blogService.addBlog(newBlog, this.props.user.token)
        addedBlog.id = addedBlog._id
        this.props.addBlog(addedBlog)
        this.setState({ newTitle: '' })
        this.setState({ newAuthor: '' })
        this.setState({ newUrl: '' })
        this.blogForm.toggleVisibility()
        this.props.notify(`a new blog ${newBlog.title} by ${newBlog.author} added`, 'info', 5)
      }
      catch (exception) {
        this.props.notify('Error creating a new blog', 'error', 5)
      }
    }
  }

  render() {
    return (
      <div>
        <Notification message='' type='info' />
        {this.props.blogs
          .sort((a, b) => a.likes > b.likes ? -1 : (a.likes < b.likes ? 1 : 0))
          .map(blog =>
            <Blog
              key={blog.id}
              blog={blog}
            />
          )}
        <h2>Create new</h2>
        <Togglable buttonLabel="new blog" ref={component => this.blogForm = component}>
          <BlogForm
            tila={this.state}
            handleTitleChange={this.handleTitleChange()}
            handleAuthorChange={this.handleAuthorChange()}
            handleUrlChange={this.handleUrlChange()}
            handleCreate={this.handleCreate()}
          ></BlogForm>
        </Togglable>
      </div>
    )
  }
}

Blogs.propTypes = {
  user: PropTypes.object.isRequired,
  blogs: PropTypes.array.isRequired
}

export default connect(null, { notify })(Blogs)