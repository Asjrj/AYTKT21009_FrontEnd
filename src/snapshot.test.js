import React from 'react'
import renderer from 'react-test-renderer'
import { User } from './containers/User'
import store from './store'

describe.only('<User />', () => {

  const theUser = {
    _id: '456',
    id: '456',
    name: 'Kalle Käyttäjä'
  }
  const blog1 = {
    id: '123',
    title: 'Snapshot otsikko',
    author: 'Snapshot kirjoittaja',
    url: 'https://reactpatterns.com/',
    likes: 1,
    user: theUser
  }
  const blog2 = {
    id: '777',
    title: 'Uusi otsikko',
    author: 'Uusi Kirjoittaja',
    url: 'https://jokumuu.com/',
    likes: 2,
    user: theUser
  }
  const blogs = [blog1, blog2]
  const users = [theUser]

  const tree = renderer
    .create(<User store={store}
      userId='456'
      blogs={blogs}
      users={users}
    />)

  test('renders correctly', () => {
    expect(tree.toJSON()).toMatchSnapshot()
  })

})