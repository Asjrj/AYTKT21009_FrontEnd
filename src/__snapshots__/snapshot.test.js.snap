// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User /> renders correctly 1`] = `
<div>
  <h2>
    Kalle Käyttäjä
  </h2>
  <h3>
    Added blogs
  </h3>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
    >
      Snapshot otsikko
       by 
      Snapshot kirjoittaja
    </li>
    <li
      className="list-group-item"
    >
      Uusi otsikko
       by 
      Uusi Kirjoittaja
    </li>
  </ul>
</div>
`;
